pkg/virt-api/rest/instancetype_test.go

```diff
@@ -0,0 +1,255 @@
+package rest
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+
+	"kubevirt.io/kubevirt/pkg/instancetype"
+
+	"github.com/emicklei/go-restful"
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	k8sfield "k8s.io/apimachinery/pkg/util/validation/field"
+
+	kubevirtcore "kubevirt.io/api/core"
+	v1 "kubevirt.io/api/core/v1"
+	instancetypev1alpha1 "kubevirt.io/api/instancetype/v1alpha1"
+	"kubevirt.io/client-go/generated/kubevirt/clientset/versioned/fake"
+	"kubevirt.io/client-go/kubecli"
+
+	"kubevirt.io/kubevirt/pkg/testutils"
+)
+
+var _ = Describe("Instancetype expansion subresources", func() {
+	const (
+		vmName      = "test-vm"
+		vmNamespace = "test-namespace"
+	)
+
+	var (
+		vmClient            *kubecli.MockVirtualMachineInterface
+		virtClient          *kubecli.MockKubevirtClient
+		instancetypeMethods *testutils.MockInstancetypeMethods
+		app                 *SubresourceAPIApp
+
+		request  *restful.Request
+		recorder *httptest.ResponseRecorder
+		response *restful.Response
+
+		vm *v1.VirtualMachine
+	)
+
+	BeforeEach(func() {
+		ctrl := gomock.NewController(GinkgoT())
+		vmClient = kubecli.NewMockVirtualMachineInterface(ctrl)
+		virtClient = kubecli.NewMockKubevirtClient(ctrl)
+		virtClient.EXPECT().GeneratedKubeVirtClient().Return(fake.NewSimpleClientset()).AnyTimes()
+		virtClient.EXPECT().VirtualMachine(vmNamespace).Return(vmClient).AnyTimes()
+
+		instancetypeMethods = testutils.NewMockInstancetypeMethods()
+
+		app = NewSubresourceAPIApp(virtClient, 0, nil, nil)
+		app.instancetypeMethods = instancetypeMethods
+
+		request = restful.NewRequest(&http.Request{})
+		recorder = httptest.NewRecorder()
+		response = restful.NewResponse(recorder)
+		response.SetRequestAccepts(restful.MIME_JSON)
+
+		vm = &v1.VirtualMachine{
+			TypeMeta: metav1.TypeMeta{},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      vmName,
+				Namespace: vmNamespace,
+			},
+			Spec: v1.VirtualMachineSpec{
+				Template: &v1.VirtualMachineInstanceTemplateSpec{
+					Spec: v1.VirtualMachineInstanceSpec{
+						Domain: v1.DomainSpec{},
+					},
+				},
+			},
+		}
+	})
+
+	testCommonFunctionality := func(callExpandInstancetypeApi func(vm *v1.VirtualMachine) *httptest.ResponseRecorder, expectedStatusError int) {
+		It("should return unchanged VM, if no instancetype and preference is assigned", func() {
+			vm.Spec.Instancetype = nil
+
+			recorder := callExpandInstancetypeApi(vm)
+			Expect(recorder.Code).To(Equal(http.StatusOK))
+
+			responseVm := &v1.VirtualMachine{}
+			Expect(json.NewDecoder(recorder.Body).Decode(responseVm)).To(Succeed())
+			Expect(responseVm).To(Equal(vm))
+		})
+
+		It("should fail if VM points to nonexistent instancetype", func() {
```

Comment:
How valid is this with the VM approach where we've already validated the instancetype and preferences during admission?

https://github.com/kubevirt/kubevirt/blob/74781324b726a6d58799be1257a34f61e6749513/pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go#L102-L105

Reply:
Actually this is valid as we don't protect against users removing an instancetype or preference after the VM is created but before it's started.

Reply:
Yes, it is not needed, but this is a unit test. I think it should test the functionality in `instancetype.go` regardless of logic in other parts of code. And it is already written, so I would keep it.

Reply:
Yup fair, I've written up https://github.com/kubevirt/kubevirt/issues/8142 for the small race window I described above.